### PR TITLE
Add support for share contact message

### PR DESCRIPTION
**Pull request by moldovaniosif**
What does this implement/fix? Explain your changes.
---------------------------------------------------
It add support for share contact message. It implements UI for the share contact cell.

Does this close any currently open issues?
------------------------------------------
NO

Where has this been tested?
---------------------------
**Devices/Simulators:** … iPhone XS Simulator, iPhone XS device

**iOS Version:** …. 12.1.4

**Swift Version:** …. 4.2

**MessageKit Version:** …. 3.0.0 beta

Share contact cell UI:
Simulator Screen Shot - 1_iPhone Xs 12 1 - 2019-03-12 at 15 11 10

### CODE REVIEW

**Comment by moldovaniosif** [comment]
I don't know why CircleCi it complains about not founding MessageInputBar.

**Comment by nathantannar4** [comment]
Ya ignore circle for now I haven't set it up right for the dependency changes. This looks promising! Will for sure include in 3.0. Busy right now but I'll take a look soon 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

File: Sources/Extensions/UIImage+Extension.swift [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ SOFTWARE.
+ */
+
+public enum ImageTye: String {
```
Comment (nathantannar4) [inline Sources/Extensions/UIImage+Extension.swift]: Typo here?

File: Sources/Protocols/ContactItem.swift [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+    func getInitials() -> String {
+        var initials = getFirstLeterFrom(firstName) + getFirstLeterFrom(lastName)
+        if initials.count == 0 {
+            initials = "#"
+        }
+        return initials
+    }
+    
+    private func getFirstLeterFrom(_ string: String?) -> String {
```
Comment (nathantannar4) [inline Sources/Protocols/ContactItem.swift]: I think swift has a built in function for this, `.prefix(_length: Int)`

File: Sources/Protocols/ContactItem.swift [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    
+    /// contact emails
+    var emails: [String] { get }
+    
+    /// Return full name from given contact item
+    /// By default the priority in computing contact name is:
+    /// 1. first + last name
+    /// 2. phone number
+    /// 3. email address
+    func getName() -> String
```
Comment (nathantannar4) [inline Sources/Protocols/ContactItem.swift]: Based on what this is returning, I think `var displayText: String { get }` would be more appropriate

File: Sources/Protocols/ContactItem.swift [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    
+    /// Return name initials from given contact. If first and last name is not set it returns #
+    func getInitials() -> String
```
Comment (nathantannar4) [inline Sources/Protocols/ContactItem.swift]: This methods use seems more like an extension to get the first char of first/last name, I think an extension on the `ContactItem` protocol with this method being public would be more appropriate rather than defining it in the protocol itself

File: Sources/Protocols/ContactItem.swift [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+/// Default Implementation for contact name and initials
+public extension ContactItem {
+
+    func getName() -> String {
```
Comment (nathantannar4) [inline Sources/Protocols/ContactItem.swift]: Instead of `String.count` I think String.isEmpty` would read better

File: Sources/Views/Cells/AudioMessageCell.swift [hunk 6]
```diff
@@ -28,32 +28,26 @@ import AVFoundation
 /// A subclass of `MessageContentCell` used to display video and audio messages.
 open class AudioMessageCell: MessageContentCell {
 
-    /// The `ImageName` enum holds the names of default iamges used to decorate play button
```
Comment (nathantannar4) [inline Sources/Views/Cells/AudioMessageCell.swift:31]: These changes seem outside the scope of this PR

File: Sources/Views/Cells/ContactMessageCell.swift [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ SOFTWARE.
+ */
+
+import UIKit
+
+open class ContactMessageCell: MessageContentCell {
+    
+    public enum ConstrainsID: String {
```
Comment (nathantannar4) [inline Sources/Views/Cells/ContactMessageCell.swift]: Typo?

File: Sources/Views/Cells/ContactMessageCell.swift [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+        case initialsContainerLeftConstraint
+        case disclouserRigtConstraint
+    }
+    
+    /// The view container that holds contact initials
+    open lazy var initialsContainerView: UIView = {
```
Comment (nathantannar4) [inline Sources/Views/Cells/ContactMessageCell.swift]: Should be public, `open` doesn't do anything as it cannot be overridden

File: Sources/Views/Cells/ContactMessageCell.swift [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+        initialsContainer.layer.shadowColor = UIColor.gray.cgColor
+        initialsContainer.layer.shadowOffset = CGSize(width: 0, height: 2)
+        initialsContainer.layer.shadowRadius = 2
+        initialsContainer.layer.shadowOpacity = 0.6
+        
+        return initialsContainer
+    }()
+    
+    /// The label that display the contact initials
+    open lazy var initialsLabel: UILabel = {
```
Comment (nathantannar4) [inline Sources/Views/Cells/ContactMessageCell.swift]: Same

File: Sources/Views/Cells/ContactMessageCell.swift [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+    /// The label that display contact name
+    open lazy var nameLabel: UILabel = {
+        let nameLabel = UILabel(frame: CGRect.zero)
+        nameLabel.numberOfLines = 0
+        
+        return nameLabel
+    }()
+    
+    /// The disclouser image view
+    open lazy var disclouserImageView: UIImageView = {
```
Comment (nathantannar4) [inline Sources/Views/Cells/ContactMessageCell.swift]: same

File: Sources/Views/Cells/ContactMessageCell.swift [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+        let initialsLabel = UILabel(frame: CGRect.zero)
+        initialsLabel.textAlignment = .center
+        initialsLabel.textColor = .darkText
+        initialsLabel.font = UIFont.preferredFont(forTextStyle: .footnote)
+
+        return initialsLabel
+    }()
+    
+    /// The label that display contact name
+    open lazy var nameLabel: UILabel = {
```
Comment (nathantannar4) [inline Sources/Views/Cells/ContactMessageCell.swift]: same

File: Sources/Views/Cells/ContactMessageCell.swift [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+        let nameLabel = UILabel(frame: CGRect.zero)
+        nameLabel.numberOfLines = 0
+
```
Comment (nathantannar4) [inline Sources/Views/Cells/ContactMessageCell.swift]: Remove extra line

File: Sources/Views/Cells/ContactMessageCell.swift [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+        nameLabel.numberOfLines = 0
+        
+        return nameLabel
+    }()
+    
+    /// The disclouser image view
+    open lazy var disclouserImageView: UIImageView = {
+        let disclouserImage = UIImage.messageKitImageWith(type: .disclouser)?.withRenderingMode(.alwaysTemplate)
+        let disclouser = UIImageView(image: disclouserImage)
+
```
Comment (nathantannar4) [inline Sources/Views/Cells/ContactMessageCell.swift]: extra line

File: Sources/Views/Cells/ContactMessageCell.swift [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+        let disclouserImage = UIImage.messageKitImageWith(type: .disclouser)?.withRenderingMode(.alwaysTemplate)
+        let disclouser = UIImageView(image: disclouserImage)
+        
+        return disclouser
+    }()
+    
+    // MARK: - Methods
+    open override func apply(_ layoutAttributes: UICollectionViewLayoutAttributes) {
+        super.apply(layoutAttributes)
+        if let attributes = layoutAttributes as? MessagesCollectionViewLayoutAttributes {
```
Comment (nathantannar4) [inline Sources/Views/Cells/ContactMessageCell.swift]: Prefer `guard let` here for consistency

File: Sources/Views/Cells/ContactMessageCell.swift [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+        }
+    }
+
+    open override func setupSubviews() {
+        super.setupSubviews()
+        messageContainerView.addSubview(initialsContainerView)
+        messageContainerView.addSubview(nameLabel)
+        messageContainerView.addSubview(disclouserImageView)
+        initialsContainerView.addSubview(initialsLabel)
+
```
Comment (nathantannar4) [inline Sources/Views/Cells/ContactMessageCell.swift]: Please remove any non-needed new-lines

File: Sources/Views/Cells/ContactMessageCell.swift [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+    
+    open func setupConstraints() {
+        initialsContainerView.constraint(equalTo: CGSize(width: 26, height: 26))
+        let initialsConstraints = initialsContainerView.addConstraints(left: messageContainerView.leftAnchor, centerY: messageContainerView.centerYAnchor,
+                                                        leftConstant: 5)
+        initialsConstraints.first?.identifier = ConstrainsID.initialsContainerLeftConstraint.rawValue
+        initialsContainerView.layer.cornerRadius = 13
+        initialsLabel.fillSuperview()
+        disclouserImageView.constraint(equalTo: CGSize(width: 20, height: 20))
+        let disclouserConstraints = disclouserImageView.addConstraints(right: messageContainerView.rightAnchor, centerY: messageContainerView.centerYAnchor,
```
Comment (nathantannar4) [inline Sources/Views/Cells/ContactMessageCell.swift]: typo here?

**Review by nathantannar4 — COMMENTED** [review]
Just looked over the code. Haven't tested yet.

File: Sources/Views/Cells/ContactMessageCell.swift [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+    }
+    
+    /// The view container that holds contact initials
+    open lazy var initialsContainerView: UIView = {
+        let initialsContainer = UIView(frame: CGRect.zero)
+        initialsContainer.backgroundColor = .white
+        initialsContainer.layer.shadowColor = UIColor.gray.cgColor
+        initialsContainer.layer.shadowOffset = CGSize(width: 0, height: 2)
+        initialsContainer.layer.shadowRadius = 2
+        initialsContainer.layer.shadowOpacity = 0.6
```
Comment (nathantannar4) [inline Sources/Views/Cells/ContactMessageCell.swift]: I think we should remove the shadow as none of our other base elements have shadow. Perhaps a `configureInitialsView` function of sorts is needed so this could customized

File: Sources/Views/Cells/AudioMessageCell.swift (hunk 6 above)
Comment (moldovaniosif) [reply Sources/Views/Cells/AudioMessageCell.swift:31]: Not quite, the `AudioMessageCell` expose a method to get a image resource with in framework by a given name. I had extracted this method in a separate extension `UIImage+Extension`. This way both classes `AudioMessageCell` and `ContactMessageCell` can easily access resources.

**Comment by nathantannar4** [comment]
@moldovaniosif You have some build errors with the functions you are using to setup constraints

File: Cartfile.resolved [hunk 18]
```diff
@@ -1,3 +1,3 @@
-github "Quick/Nimble" "v7.3.4"
-github "Quick/Quick" "v1.3.4"
-github "nathantannar4/InputBarAccessoryView" "4.2.1"
+github "MessageKit/MessageInputBar" "0.4.0"
```
Comment (nathantannar4) [inline Cartfile.resolved]: Please undo this change, I am changing the dependency

File: Sources/Extensions/UIView+Extensions.swift [hunk 19]
```diff
@@ -97,18 +99,6 @@ extension UIView {
...
-            let constraint = centerYAnchor.constraint(equalTo: centerY, constant: centerYConstant)
-            constraint.identifier = "centerY"
-            constraints.append(constraint)
-        }
-
-        if let centerX = centerX {
-            let constraint = centerXAnchor.constraint(equalTo: centerX, constant: centerXConstant)
-            constraint.identifier = "centerX"
-            constraints.append(constraint)
-        }
```
Comment (nathantannar4) [inline Sources/Extensions/UIView+Extensions.swift]: Any reason these were removed? Its what is causing the build error

File: Sources/Models/Sender.swift [hunk 20]
```diff
@@ -25,33 +25,33 @@
...
-    }
+    public let id: String
 
     /// The display name of a sender.
     public let displayName: String
 
     // MARK: - Intializers
 
-    public init(senderId: String, displayName: String) {
-        self.senderId = senderId
```
Comment (nathantannar4) [inline Sources/Models/Sender.swift]: Please revert all these changes to `Sender`, outside scope of PR and breaks Sender from being SenderType

File: Sources/Protocols/ContactItem.swift [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+import Foundation
+
+/// A protocol used to represent the data for a contact message.
+public protocol ContactItem {
+    
+    /// contact first name
+    var firstName: String? { get }
+    
+    /// contact last name
+    var lastName: String? { get }
```
Comment (nathantannar4) [inline Sources/Protocols/ContactItem.swift]: I think it would be better to merge firstName and lastName into one, `displayName`. A contact may be a company and thus first/last name isn't appropriate

File: Sources/Views/Cells/AudioMessageCell.swift [hunk 22]
```diff
@@ -28,32 +28,26 @@ import AVFoundation
...
+        let playImage = UIImage.messageKitImageWith(type: .play)
+        let pauseImage = UIImage.messageKitImageWith(type: .pause)
         playButton.setImage(playImage?.withRenderingMode(.alwaysTemplate), for: .normal)
         playButton.setImage(pauseImage?.withRenderingMode(.alwaysTemplate), for: .selected)
         return playButton
     }()
 
     /// The time duration lable to display on audio messages.
-    public lazy var durationLabel: UILabel = {
+    open lazy var durationLabel: UILabel = {
```
Comment (nathantannar4) [inline Sources/Views/Cells/AudioMessageCell.swift]: Please revert to public

File: Sources/Views/Cells/AudioMessageCell.swift [hunk 23]
```diff
@@ -28,32 +28,26 @@ import AVFoundation
...
         let durationLabel = UILabel(frame: CGRect.zero)
         durationLabel.textAlignment = .right
         durationLabel.font = UIFont.systemFont(ofSize: 14)
         durationLabel.text = "0:00"
         return durationLabel
     }()
 
-    public lazy var progressView: UIProgressView = {
+    open lazy var progressView: UIProgressView = {
```
Comment (nathantannar4) [inline Sources/Views/Cells/AudioMessageCell.swift]: Please revert to public

**Review by nathantannar4 — COMMENTED** [review]
Please be careful merging in other branches to yours. It looks like you merged in the master branch (or started from) and this thus reverts changes made in the 3.0.0-beta branch

**Comment by moldovaniosif** [comment]
The merge with mater was a mistake, I reverted those commits.

File: Sources/Protocols/ContactItem.swift [hunk 24]
```diff
@@ -27,47 +27,15 @@ import Foundation
...
     
-    /// contact first name
-    var firstName: String? { get }
+    /// contact displayed name
+    var displayName: String { get }
     
-    /// contact last name
-    var lastName: String? { get }
+    /// initials from contact first and last name
+    var initials: String { get }
```
Comment (nathantannar4) [inline Sources/Protocols/ContactItem.swift:34]: Nice! I like this better

**Review by nathantannar4 — APPROVED** [review]
